File: golang/binary-tree-bfs/max_level_sum_binary_tree.go
package main

func maxLevelSum(root *TreeNode) int {
	if root == nil {
		return 0
	}

	level := 1
	maxLevel := 1
	maxSum := root.Val

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		levelSum := 0

		nextQueue := []*TreeNode{}

		for _, node := range queue {
			levelSum += node.Val

			if node.Left != nil {
				nextQueue = append(nextQueue, node.Left)
			}

			if node.Right != nil {
				nextQueue = append(nextQueue, node.Right)
			}
		}

		if levelSum > maxSum {
			maxSum = levelSum
			maxLevel = level
		}

		queue = nextQueue

		level++
	}

	return maxLevel
}

File: java/src/main/java/com/sanedge/interview/binary_tree_bfs/MaxLevelSumBinaryTree.java
package com.sanedge.interview.binary_tree_bfs;

import com.sanedge.interview.utils.TreeNode;
import java.util.LinkedList;
import java.util.Queue;

public class MaxLevelSumBinaryTree {

    public int maxLevelSum(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int level = 1;
        int maxLevel = 1;
        int maxSum = root.val;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSum = 0;
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();

                levelSum += node.val;

                if (node.left != null) {
                    queue.offer(node.left);
                }

                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            if (levelSum > maxSum) {
                maxSum = levelSum;
                maxLevel = level;
            }

            level++;
        }

        return maxLevel;
    }
}
